caas/kubernetes/provider/events.go

```diff
@@ -66,3 +67,18 @@ func (k *kubernetesClient) watchEvents(objName string, objKind string) (watcher.
 	)
 	return k.newWatcher(factory.Core().V1().Events().Informer(), objName, k.clock)
 }
+
+var imageNotFoundRegex = regexp.MustCompile(`Failed to pull image "(\S+)": rpc error: code = NotFound .*`)
+
+// isImageNotFound determines if the given event is an "image not found" error,
+// and if so, returns the missing image name.
+func isImageNotFound(evt core.Event) (ok bool, imageTag string) {
+	if evt.Type != core.EventTypeWarning {
+		return false, ""
+	}
+	match := imageNotFoundRegex.FindStringSubmatch(evt.Message)
```

Comment:
According to the Kubernetes go docs message is a human readable string that offers no guarantees around versioning and change. This will break on us in a big way.

There is a Reason field that is the machine readable variant.

Reply:
in general, I agree with you - however, in this specific case, this is the event struct that Kubernetes provides us
```
v1.Event{
  Type: "Warning",
  Reason: "Failed",
  Message: "Failed to pull image \"jujusolutions/jujud-operator:3.1-fake1\": rpc error: code = NotFound desc = failed to pull and unpack image \"docker.io/jujusolutions/jujud-operator:3.1-fake1\": failed to unpack image on snapshotter overlayfs: unexpected media type text/html for sha256:b6a78ff088000afc609fcbc701d18704ddb944e867af0dadd520d4bf0e5af328: not found",
  ...
}
```

As you can see, the `Type` and `Reason` fields here provide no information beyond "something went wrong". So we have no choice here but to parse the message in order to work out why the pod is failing.